File: android/app/src/main/java/org/moon/moonmeet/notification/NotificationServiceExtension.java
/*
 * This is the source code of Moon Meet CrossPlatform.
 * It is licensed under GNU GPL v. 3.
 * You should have received a copy of the license in this archive (see LICENSE).
 *
 * Copyright Rayen sbai, 2021-2022.
 */

package org.moon.moonmeet.notification;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import androidx.core.app.NotificationCompat;

import com.onesignal.OSMutableNotification;
import com.onesignal.OSNotification;
import com.onesignal.OSNotificationReceivedEvent;
import com.onesignal.OneSignal.OSRemoteNotificationReceivedHandler;

import org.json.JSONException;
import org.json.JSONObject;
import org.moon.moonmeet.R;

import java.io.IOException;
import java.io.InputStream;
import java.math.BigInteger;
import java.net.HttpURLConnection;
import java.net.URL;

@SuppressWarnings("unused")
public class NotificationServiceExtension implements OSRemoteNotificationReceivedHandler {
    private static final String TAG = "NotificationServiceExtension";

    private Bitmap getBitmapFromURL(String src) {
        try {
            URL url = new URL(src);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setDoInput(true);
            connection.connect();
            InputStream input = connection.getInputStream();
            return BitmapFactory.decodeStream(input);
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public void remoteNotificationReceived(Context context, OSNotificationReceivedEvent notificationReceivedEvent) {
        OSNotification osNotification = notificationReceivedEvent.getNotification();
        JSONObject NotificationAdditionalData = osNotification.getAdditionalData();
        Log.i(TAG, NotificationAdditionalData.toString());
        OSMutableNotification mutableNotification = osNotification.mutableCopy();
        try {
            if (NotificationAdditionalData.getString("type").equals("chat")) {
                mutableNotification.setExtender(builder -> {
                    try {
                        builder.setChannelId("messages");
                        builder.setSmallIcon(R.drawable.moon_icon);
                        builder.setStyle(new NotificationCompat.BigTextStyle().bigText(NotificationAdditionalData.getString("messageDelivered")));
                        builder.setColor(new BigInteger("FF566193", 16).intValue());
                        builder.setContentTitle(NotificationAdditionalData.getString("senderName"));
                        builder.setContentText("New message from " + NotificationAdditionalData.getString("senderName"));
                        builder.setPriority(NotificationCompat.PRIORITY_HIGH);
                        builder.setWhen(Long.parseLong(NotificationAdditionalData.getString("messageTime")));
                        builder.setLargeIcon(getBitmapFromURL(NotificationAdditionalData.getString("senderPhoto")));
                    } catch (JSONException e) {
                        e.printStackTrace();
                        notificationReceivedEvent.complete(null);
                    }
                    Log.i(TAG, "returning a builder from notification chat type.");
                    return builder;
                });
            } else {
                mutableNotification.setExtender(builder -> {
                    builder.setChannelId("default");
                    builder.setSmallIcon(R.drawable.moon_icon);
                    builder.setColor(new BigInteger("FF566193", 16).intValue());
                    builder.setContentTitle(osNotification.getTitle());
                    builder.setContentText(osNotification.getBody());
                    return builder;
                });
            }
            notificationReceivedEvent.complete(mutableNotification);
        } catch (JSONException e) {
            e.printStackTrace();
            notificationReceivedEvent.complete(null);
        }
    }
}

File: android/app/src/main/java/org/moon/moonmeet/notification/NotificationServiceExtension.kt
/*
 * This is the source code of Moon Meet CrossPlatform.
 * It is licensed under GNU GPL v. 3.
 * You should have received a copy of the license in this archive (see LICENSE).
 *
 * Copyright Rayen Sbai, 2021-2024.
 */
package org.moon.moonmeet.notification

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import androidx.annotation.Keep
import androidx.core.app.NotificationCompat
import com.onesignal.notifications.IDisplayableMutableNotification
import com.onesignal.notifications.INotificationReceivedEvent
import com.onesignal.notifications.INotificationServiceExtension
import org.json.JSONException
import org.json.JSONObject
import org.moon.moonmeet.R
import java.io.IOException
import java.math.BigInteger
import java.net.HttpURLConnection
import java.net.URL

@Keep
class NotificationServiceExtension : INotificationServiceExtension {

    private fun getBitmapFromURL(src: String): Bitmap? {
        return try {
            val url = URL(src)
            val connection = url.openConnection() as HttpURLConnection
            connection.doInput = true
            connection.connect()
            val input = connection.inputStream
            BitmapFactory.decodeStream(input)
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }

    override fun onNotificationReceived(event: INotificationReceivedEvent) {
        val notification: IDisplayableMutableNotification = event.notification
        val notificationAdditionalData: JSONObject? = notification.additionalData
        Log.i(TAG, notificationAdditionalData.toString())

        try {
            if (notificationAdditionalData != null) {
                if (notificationAdditionalData.getString("type") == "chat") {
                    notification.setExtender { builder: NotificationCompat.Builder ->
                        if (notificationAdditionalData.has("imageDelivered")) {
                            try {
                                builder.setChannelId("messages")
                                builder.setSmallIcon(R.drawable.moon_icon)
                                builder.setColor(BigInteger("FF566193", 16).toInt())
                                builder.setContentTitle(notificationAdditionalData.getString("senderName"))
                                builder.setContentText(notificationAdditionalData.getString("imageDelivered"))
                                builder.setPriority(NotificationCompat.PRIORITY_HIGH)
                                builder.setWhen(
                                    notificationAdditionalData.getString("messageTime").toLong()
                                )
                                builder.setLargeIcon(
                                    getBitmapFromURL(notificationAdditionalData.getString("senderPhoto"))
                                )
                            } catch (e: JSONException) {
                                e.printStackTrace()
                                event.preventDefault()
                            }
                        } else if (notificationAdditionalData.has("messageDelivered")) {
                            try {
                                builder.setChannelId("messages")
                                builder.setSmallIcon(R.drawable.moon_icon)
                                builder.setStyle(
                                    NotificationCompat.BigTextStyle()
                                        .bigText(notificationAdditionalData.getString("messageDelivered"))
                                )
                                builder.setColor(BigInteger("FF566193", 16).toInt())
                                builder.setContentTitle(notificationAdditionalData.getString("senderName"))
                                builder.setContentText(
                                    "New message from " + notificationAdditionalData.getString("senderName")
                                )
                                builder.setPriority(NotificationCompat.PRIORITY_HIGH)
                                builder.setWhen(
                                    notificationAdditionalData.getString("messageTime").toLong()
                                )
                                builder.setLargeIcon(
                                    getBitmapFromURL(notificationAdditionalData.getString("senderPhoto"))
                                )
                            } catch (e: JSONException) {
                                e.printStackTrace()
                                event.preventDefault()
                            }
                        }
                        builder
                    }
                } else {
                    notification.setExtender { builder: NotificationCompat.Builder ->
                        builder.setChannelId("default")
                        builder.setSmallIcon(R.drawable.moon_icon)
                        builder.setColor(BigInteger("FF566193", 16).toInt())
                        builder.setContentTitle(notification.title)
                        builder.setContentText(notification.body)
                        builder
                    }
                }
            }
        } catch (e: JSONException) {
            e.printStackTrace()
            event.preventDefault()
        }
    }

    companion object {
        private const val TAG = "NotificationServiceExtension"
    }
}
